### PR TITLE
Add docs and participants support for blocked and allowed email domains feature.

### LINKED ISSUE #1606: Allow and block particular email domains for a challenge
Currently, anyone can participate in a challenge. It would be great to have a functionality to allow/block only particular email domains to participate in a challenge as specified by the challenge host. 

## Deliverables

- [x] Add a field to Challenge model to store the `allowed_email_domains` (default=None, blank=True)
- [x] Add a field to Challenge model to store the `blocked_emails_domains` (default=None, blank=True)
- [ ] Update the challenge creation API to add support to incorporate `allowed_email_domains` and `blocked_email_domains`
- [ ] Update the Challenge creation using zip file documentation to add support for `allowed_email_domains`   and `blocked_email_domains`
- [ ] Update the challenge participation API to only allow `allowed_email_domains` and block `blocked_email_domains`

### PR DESCRIPTION
**Pull request by isht3**
Docs.
* **allowed_email_domains**: A list of domains allowed to participate in the challenge. Leave blank if everyone is allowed to participate. (e.g. `["domain1.com", "domain2.org", "domain3.in"]` Participants with these email domains will only be allowed to participate.)
 
 * **blocked_emails_domains**: A list of domains not allowed to participate in the challenge. Leave blank if everyone is allowed to participate. (e.g. `["domain1.com", "domain2.org", "domain3.in"]` The participants with these email domains will not be allowed to participate.)

In blocked list.

Not in allowed list.

Fixes #1606 #

### CODE REVIEW

**Comment by deshraj** [comment]
I think you need to add the support in the challenge creation API as well. Just adding the documentation doesn't make sense.

**Comment by isht3** [comment]
@deshraj I did. It was in the PR #1612 :smile:

**Comment by deshraj** [comment]
Changing the serializer fields will not make sure that challenge hosts can add the list of blocked emails, allowed emails. They have to specify it somewhere and updating the challenge creation API to take these two as input is the way to go. So, 

1. you have to add these two fields in the yaml file for challenge creation
2. Add support for these two fields in the challenge creation API
3. And hence you have to modify the documentation for challenge creation using YAML file.

**Comment by isht3** [comment]
@deshraj I'm not quite sure I understand. Because when I tried testing the Challenge Creation, it worked.

When I gave the following configuration in the YAML file.



I got this in the admin panel.



Which I think is the desired behavior. Am I missing something important?

**Comment by deshraj** [comment]
Yeah, you got it right. But there are some changes that you have to do:

1. The email domains have to be stored in the format `@cloudcv.org` instead of just storing the name like `cloudcv` etc.

2. Make sure that you store all the domain names in lowercase. 

3. In the above example, it should be the case that if you try to participate in the created challenge using any email id that has `yahoo.com`/`gmail.com`/`outlook.com` should not be allowed to participate in the challenge and is shown a message on the UI.

**Comment by isht3** [comment]
@deshraj The lowercase checks and the "@" would be appended while during the check. I thought the user would be having more flexibility because of that. 

Is checking for `@cloudcv.org` instead of `@cloudcv` going to be more effective? 

I just added the final change along with the tests. :smile:

**Comment by deshraj** [comment]
> @deshraj The lowercase checks and the "@" would be appended while during the check. I thought the user would be having more flexibility because of that.

Agree. We can just drop the `@` from what I said.

> Is checking for @cloudcv.org instead of @cloudcv going to be more effective?

Yes because similar domains such as cloudcv can have many extensions such as `cloudcv.org` or `cloudcv.com` etc.

**Comment by isht3** [comment]
@deshraj Yes! I changed it. :smile:

File: apps/challenges/views.py [hunk 1]
```diff
@@ -168,6 +168,39 @@ def add_participant_team_to_challenge(request, challenge_pk, participant_team_pk
...
+            domain = "@" + domain
+            if domain.lower() in user_email.lower():
+                present = True
+                break
+        if not present:
+            message = 'Sorry, only users with {} domain is allowed to participate in this challenge.'
+            domains = ""
+            for domain in challenge.allowed_email_domains:
+                domains = "{}{}{}".format(domains, "/", domain)
+            domains = domains[1:]
```
[the remark this answers is not in the document]
Comment (isht3) [reply apps/challenges/views.py]: @deshraj This is basically for pretty printing. It prints the domains like `gmail.com/yahoo.com/outlook.com` like in the screenshot above.

[the remark this answers is not in the document]
Comment (deshraj) [reply apps/challenges/views.py]: Yeah but why are we starting from the second index and not the first index of `domains` variable?

[the remark this answers is not in the document]
Comment (deshraj) [reply apps/challenges/views.py]: Oh ok. Got it. It's just to remove the first `/`

[the remark this answers is not in the document]
Comment (deshraj) [reply apps/challenges/views.py]: Looks good overall. Please incorporate the other requested changes and then we can merge and deploy this feature.

**Comment by isht3** [comment]
@deshraj I made the required changes.

**Comment by deshraj** [comment]
@isht3 please fix the travis build issue.

**Comment by isht3** [comment]
@deshraj Fixed!

**Pull request merged**